interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,426 @@
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    interface=###DBUS_INTERFACE###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
+# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
+# connected snaps.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
+const dbusConnectedPlugSecComp = `
+getsockname
+recvmsg
+sendmsg
+sendto
+`
+
+type DbusInterface struct{}
+
+func (iface *DbusInterface) Name() string {
+	return "dbus"
+}
+
+// Obtain yaml-specified bus well-known name
+func (iface *DbusInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
+	// bus attribute
+	bus, ok := attribs["bus"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'bus'")
+	}
+
+	if bus != "session" && bus != "system" {
+		return "", "", fmt.Errorf("bus '%s' must be one of 'session' or 'system'", bus)
+	}
+
+	// name attribute
+	name, ok := attribs["name"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'name'")
+	}
+
+	err := interfaces.ValidateDBusBusName(name)
+	if err != nil {
+		return "", "", err
+	}
+
+	// snapd has AppArmor rules (see above) allowing binds to busName-PID
+	// so to avoid overlap with different snaps (eg, busName running as PID
+	// 123 and busName-123), don't allow busName to end with -PID. If that
+	// rule is removed, this limitation can be lifted.
+	invalidSnappyBusName := regexp.MustCompile("-[0-9]+$")
+	if invalidSnappyBusName.MatchString(name) {
+		return "", "", fmt.Errorf("DBus bus name must not end with -NUMBER")
+	}
+
+	return bus, name, nil
+}
+
+// Determine AppArmor dbus abstraction to use based on bus
+func getAppArmorAbstraction(bus string) (string, error) {
+	var abstraction string
+	if bus == "system" {
+		abstraction = "dbus-strict"
+	} else if bus == "session" {
+		abstraction = "dbus-session-strict"
+	} else {
+		return "", fmt.Errorf("unknown abstraction for specified bus '%q'", bus)
+	}
+	return abstraction, nil
+}
+
+// Calculate individual snippet policy based on bus and name
+func getAppArmorSnippet(policy []byte, bus string, name string) []byte {
+	old := []byte("###DBUS_BUS###")
+	new := []byte(bus)
+	snippet := bytes.Replace(policy, old, new, -1)
+
+	old = []byte("###DBUS_NAME###")
+	new = []byte(name)
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path (eg 'org.foo' to '/org/foo{,/**}')
+	var pathBuf bytes.Buffer
+	pathBuf.WriteString(`"/`)
+	pathBuf.WriteString(strings.Replace(name, ".", "/", -1))
+	pathBuf.WriteString(`{,/**}"`)
+
+	old = []byte("###DBUS_PATH###")
+	new = pathBuf.Bytes()
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus interface (eg, 'org.foo' to 'org.foo{,.*}')
+	var ifaceBuf bytes.Buffer
+	ifaceBuf.WriteString(`"`)
+	ifaceBuf.WriteString(name)
+	ifaceBuf.WriteString(`{,.*}"`)
```

Comment:
As mentioned on the mailing list: https://lists.ubuntu.com/archives/snapcraft/2016-December/001998.html this initial implementation of the interface intentionally does not address complex dbus services. It is meant to address leaf applications (eg, GNOME and KDE) that bind to a well-known name.

I'd prefer not to delay this PR any longer by adding more functionality to it. @niemeyer and I agreed in this PR that we would implement what we have and then see how people are using it and potentially add DBus paths, interfaces, activation, complex bus policy, etc down the line. Due keep in mind that generic interfaces like this one are a new concept in snappy that is in some ways counter to the initial goals of interfaces, and we want to make sure we design and implement them in a controlled fashion.

As for storage framework in particular, if you haven't already, I suggest reading the aforementioned thread. At this time, it will need to be its own interface (and that's ok-- it's easy getting interfaces into snappy :). Perhaps at a future date the dbus interface will get to the point to support all DBus services generically, but not just yet.